### ISSUE TITLE
FR: Allow toolbar button to toggle extension functionality on/off

### ISSUE THREAD

**Issue by Gitoffthelawn** [airtower-luna/referer-mod#41]
The fun new toolbar button opens the settings page.  This is certainly reasonable, but I find I want to toggle the extension's functionality on/off more frequently than I want to change any of its settings.  I frequently need to momentarily toggle the extension off to help determine what is causing a broken page to not load as expected.

Would there be a way to allow the toolbar button to toggle the extension's functionality, and have access to its options page via the toolbar button's context menu?

**Comment by airtower-luna** [comment]
There is the option to have the button open a popup instead of directly opening the settings page, and that could then e.g. have buttons to open settings or enable/disable. It's something I'm considering, not sure which way would be best yet:

* Disable globally. I don't really like that, because experimenting in one tab shouldn't affect requests from other tabs that happen meanwhile.
* Disable for the current tab. Wouldn't affect sites in other tabs, but might not be ideal for people who open the same site in multiple tabs, sites that open popups, or people who then open an unrelated site in the same tab.
* Disable for the current domain. The way I imagine this it'd check the domain of the site open in a tab, so it'd affect e.g. iframes from different domains, too. This is similar to how tools like uBlock Origin handle settings, so it might be most intuitive.

Any opinions on which would be best? Adding more than one might be an option, too, but I don't want to overload the UI.

**Comment by Gitoffthelawn** [comment]
**TL;DR: I'll spend some time thinking about it. :)**

Hmmm... all good ideas.  Let me think about it for a bit (intentionally ambiguous timeframe!).

I was initially thinking of keeping it real simple with a global disable, but the issue of affecting other tabs is certainly something to ponder.  My mental model for extensions is that they affect the entire browser, so a global disable seems natural to me.

I also use (and contribute to) another wonderful extension called [Redirector](https://github.com/einaregilsson/Redirector), and when it is disabled via its options, it disables globally, and that has worked well for me and the primary developer, Einar.  To my recollection, no one has ever complained about that functionality or been surprised by it.

All that said, I think it's worth some thought to figure out a good implementation.  All three of your bullet points make good sense, and all would make good options.  But as you mention, keeping the UI from being overloaded is important, as is matching functionality with user expectations.

**Comment by airtower-luna** [comment]
I've been experimenting a bit and will go with global toggle for now. The reason is that checking the domain of the current tab from the popup requires the "tabs" permission, and the granular disable doesn't feel important enough to me right now to demand more permissions. Though looking at all HTTP requests is pretty deep already anyway. Can always add it later if needed. :sweat_smile:

**Comment by airtower-luna** [comment]
You're welcome to test #42. I want to add a test before merging, but it works already.

**Comment by Gitoffthelawn** [comment]
What do you think of having the toolbar button toggle the extension, and then a context menu for the button to access the settings?  Just as an example of popular extensions with context menuitems for the toolbar button, there's uBlock Origin. 

Redirector does the popup menu thing, and I gotta say, I'm not a big fan of that design.  No matter what you want to do, it always requires at least 2 clicks.

**Comment by airtower-luna** [comment]
A one-click toggle is actually something I specifically wanted to avoid. The reason why I placed the settings button on top of the popup and the toggle below is similar: to minimize the risk of misclicking on it. :sweat_smile:

And also because I feel I want to open the settings more often than I'd want to disable the modifications in general. :smile_cat:

**Comment by Gitoffthelawn** [comment]
My desire is the opposite, but that's okay... that's what makes us human. 🧑‍🍼 

As a bit of an aside... actually... I see you have discussions enabled, so no aside here... I'll create a discussion item. 😸

**Comment by Gitoffthelawn** [comment]
Here's the discussion I promised: https://github.com/airtower-luna/referer-mod/discussions/43  😸

**Comment by airtower-luna** [comment]
Yeah, there's a lot of personal feelings/opinion in what's convenient. Either way I'm sure this is better than having to go to the add-ons manager. :wink:

**Comment by Gitoffthelawn** [comment]
> Yeah, there's a lot of personal feelings/opinion in what's convenient. Either way I'm sure this is better than having to go to the add-ons manager. wink

The funny thing is that lately I've been going to the add-ons manager to review/edit the extension's settings, and now it takes an extra click to find them there (and then a new tab is opened).  I'm curious to see if, over the long term, the new method feels easier than what I was used to.  It will be an interesting little experiment.

BTW, is there any way to still allow the `Options` tab to appear for this extension in the add-ons manager?  Or is it mutually exclusive to have a separate browser tab and allow the `Options` tab within the add-ons manager?

**Comment by airtower-luna** [comment]
> Or is it mutually exclusive to have a separate browser tab and allow the `Options` tab within the add-ons manager?

Those are mutually exclusive if you want to use the proper "open options" API. It is possible to open the separate options tab by URL, but then multiple clicks on the settings button will open multiple tabs. :sweat_smile:

It would be possible to open the options tab in the add-on manager via the toolbar button or popup, though. Do you think that'd be more useful? My main reason to switch to a separate tab was to have more space, especially on small screens. Things like #15 will need even more, unless there's a responsive redesign first.
